File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int l =0 , r = matrix[0].size()-1,u = 0 , d = matrix.size()-1;
       vector<int>ans;
     while (l <= r && u <= d) {
        // Traverse top row
        for (int i = l; i <= r; i++) {
            ans.push_back(matrix[u][i]);
        }
        u++; // Move down the upper boundary

        // Traverse right column
        for (int i = u; i <= d; i++) {
            ans.push_back(matrix[i][r]);
        }
        r--; // Move left the right boundary

        // Traverse bottom row (only if u <= d to avoid duplicates)
        if (u <= d) {
            for (int i = r; i >= l; i--) {
                ans.push_back(matrix[d][i]);
            }
            d--; // Move up the lower boundary
        }

        // Traverse left column (only if l <= r to avoid duplicates)
        if (l <= r) {
            for (int i = d; i >= u; i--) {
                ans.push_back(matrix[i][l]);
            }
            l++; // Move right the left boundary
        }
    }
        return ans;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] mat) {
        List<Integer>ans = new ArrayList<>();
        int top =0, bottom = mat.length-1,l = 0, r = mat[0].length-1;
        while(top <= bottom && l<= r){
           for(int i=l;i<=r;i++){
             ans.add(mat[top][i]);
           }
           top++;
           for(int i=top;i<=bottom;i++){
              ans.add(mat[i][r]);
           }
           r--;
           if(top<=bottom){
           for(int i=r;i>=l;i--){
              ans.add(mat[bottom][i]);
           }
           bottom--;
           }
           
           if(l<=r){
           for(int i=bottom;i>=top;i--){
             ans.add(mat[i][l]);
           }
           l++;
           }
           
        }
        return ans;
    }
}